File: csharp/1247.0_Minimum_Swaps_to_Make_Strings_Equal.cs
/*
执行用时：60 ms, 在所有 C# 提交中击败了63.64% 的用户
内存消耗：35.7 MB, 在所有 C# 提交中击败了18.18% 的用户
通过测试用例：70 / 70
*/
public class Solution {
    public int MinimumSwap(string s1, string s2) {
        int xy = 0, yx = 0;
        int n = s1.Length;
        for (int i = 0; i < n; i++) {
            char a = s1[i], b = s2[i];
            if (a == 'x' && b == 'y') {
                xy++;
            }
            else if (a == 'y' && b == 'x') {
                yx++;
            }
        }
        if (((xy + yx) & 1) == 1) {
            return -1;
        }
        return xy / 2 + yx / 2 + xy % 2 + yx % 2;
    }
}

File: leetcode-cn/1247.0_Minimum_Swaps_to_Make_Strings_Equal.py
'''
hash table

执行用时：32 ms, 在所有 Python3 提交中击败了90.97% 的用户
内存消耗：14.7 MB, 在所有 Python3 提交中击败了100.00% 的用户
通过测试用例：70 / 70
'''
class Solution:
    def minimumSwap(self, s1: str, s2: str) -> int:
        # diff pairs count, (x, y) (y, x)
        diff = Counter()
        for a, b in zip(s1, s2):
            if a != b:
                diff[(a, b)] += 1
        xy = diff[('x', 'y')]
        yx = diff[('y', 'x')]
        if xy & 1 and yx & 1:
            return 2 + (xy // 2) + (yx // 2)
        elif not (xy & 1) and not (yx & 1):
            return (xy // 2) + (yx // 2)
        else:
            return -1

'''
执行用时：36 ms, 在所有 Python3 提交中击败了75.69% 的用户
内存消耗：14.7 MB, 在所有 Python3 提交中击败了100.00% 的用户
通过测试用例：70 / 70
'''
class Solution:
    def minimumSwap(self, s1: str, s2: str) -> int:
        xy = yx = 0
        for a, b in zip(s1, s2):
            if a == 'x' and b == 'y':
                xy += 1
            elif a == 'y' and b == 'x':
                yx += 1
        if (xy + yx) & 1:
            return -1
        return xy // 2 + yx // 2 + xy % 2 + yx % 2
        
        